packages/@aws-cdk/aws-ecs/lib/drain-hook/instance-drain-hook.ts

```diff
@@ -0,0 +1,101 @@
+import autoscaling = require('@aws-cdk/aws-autoscaling');
+import iam = require('@aws-cdk/aws-iam');
+import lambda = require('@aws-cdk/aws-lambda');
+import sns = require('@aws-cdk/aws-sns');
+import cdk = require('@aws-cdk/cdk');
+import path = require('path');
+import { ICluster } from '../cluster';
+
+// Reference for the source in this package:
+//
+// https://github.com/aws-samples/ecs-refarch-cloudformation/blob/master/infrastructure/lifecyclehook.yaml
+
+/**
+ * Properties for instance draining hook
+ */
+export interface InstanceDrainHookProps {
+  /**
+   * The AutoScalingGroup to install the instance draining hook for
+   */
+  autoScalingGroup: autoscaling.IAutoScalingGroup;
+
+  /**
+   * The cluster on which tasks have been scheduled
+   */
+  cluster: ICluster;
+
+  /**
+   * How many seconds to give tasks to drain before the instance is terminated anyway
+   *
+   * Must be between 0 and 900.
+   *
+   * @default 900
+   */
+  drainTimeSeconds?: number;
+}
+
+/**
+ * A hook to drain instances from ECS traffic before they're terminated
+ */
+export class InstanceDrainHook extends cdk.Construct {
+  constructor(parent: cdk.Construct, id: string, props: InstanceDrainHookProps) {
+    super(parent, id);
+
+    const drainTimeSeconds = props.drainTimeSeconds !== undefined ? props.drainTimeSeconds : 300;
+
+    if (drainTimeSeconds < 0 || drainTimeSeconds > 900) {
+      throw new Error(`Drain time must be between 0 and 900 seconds, got: ${drainTimeSeconds}`);
+    }
+
+    // Invoke Lambda via SNS Topic
+    const topic = new sns.Topic(this, 'Topic');
+    const fn = new lambda.Function(this, 'Function', {
+      code: lambda.Code.directory(path.join(__dirname, 'lambda-source')),
+      handler: 'index.lambda_handler',
+      runtime: lambda.Runtime.Python36,
+      // Timeout: some extra margin for additional API calls made by the Lambda,
+      // up to a maximum of 15 minutes.
+      timeout: Math.min(drainTimeSeconds + 10, 900),
+      environment: {
+        CLUSTER: props.cluster.clusterName
+      }
+    });
+
+    // Hook everything up: ASG -> Topic, Topic -> Lambda
+    props.autoScalingGroup.onLifecycleTransition('DrainHook', {
+      lifecycleTransition: autoscaling.LifecycleTransition.InstanceTerminating,
+      defaultResult: autoscaling.DefaultResult.Continue,
+      notificationTarget: topic,
+      heartbeatTimeoutSeconds: drainTimeSeconds,
+    });
+    topic.subscribeLambda(fn);
+
+    // Permissions
+    topic.grantPublish(fn.role);
```

Comment:
I think this is done automatically when you do `topic.subscribeLambda`

Reply:
No, that's the other way around. But also, this isn't necessary anymore, I rewrote the Lambda to sleep instead of republishing itself to the topic.